File: 106. Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map < int , int > pos;
    vector < int > pstOrd;
    int idx;
    TreeNode* dfs(int low, int high) {
        if( low > high ) return NULL;
        int val = pstOrd[idx--];
        TreeNode* root = new TreeNode(val);
        int p = pos[val];
       
        root->right = dfs( p + 1, high);
         root->left = dfs(low, p - 1);
        return root;
    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        pstOrd = postorder;
        idx = pstOrd.size() - 1;
        for( int i = 0 ; i < inorder.size() ; i++ ) {
            pos[inorder[i]] = i;
        }
        return dfs( 0 , idx );
        
    }
};

File: python/Lakshay/106. Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        idx = len(postorder) - 1
        inordered_pos = {}
        for index, value in enumerate(inorder):
            inordered_pos[value] = index

        def dfs(low, high):
            nonlocal idx
            if low > high:
                return None
            root_val = postorder[idx]
            idx -= 1
            root = TreeNode(root_val)

            mid = inordered_pos[root_val]
            root.right = dfs( mid + 1, high )
            root.left = dfs( low, mid - 1)
           
            return root
        return dfs( 0 , len(postorder) - 1 )  
